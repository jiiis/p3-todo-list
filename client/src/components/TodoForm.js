import React, { useState } from 'react'

const TodoForm = ({ status, onStatusChange, onSubmit }) => {
  // States
  const [description, setDescription] = useState('')

  // Event handlers
  const handleDescriptionFieldChange = (event) => {
    setDescription(event.target.value)
  }

  const handleStatusFieldChange = (event) => {
    onStatusChange(event.target.value)
  }

  const handleFormSubmit = async (event) => {
    event.preventDefault()

    if (!isDescriptionValid()) {
      return
    }

    await onSubmit(description.trim())

    setDescription('')
  }

  // Helpers
  const isDescriptionValid = () => {
    return !/^\s*$/.test(description)
  }

  return (
    <div>
      <form
        onSubmit={handleFormSubmit}
        data-testid="todo-form"
      >
        <label htmlFor="todo-description-field">Create todo item</label>
        <input
          id="todo-description-field"
          type="text"
          name="description"
          value={description}
          onChange={handleDescriptionFieldChange}
          data-testid="todo-status-description"
        />
        <button data-testid="todo-form-button">Add</button>
      </form>
      <div>
        <label htmlFor="todo-status-select">Select status</label>
        <select
          id="todo-status-select"
          name="status"
          value={status}
          onChange={handleStatusFieldChange}
          data-testid="todo-status-select"
        >
          <option value="all">All</option>
          <option value="completed">Completed</option>
          <option value="uncompleted">Uncompleted</option>
        </select>
      </div>
    </div>
  )
}

export default TodoForm
