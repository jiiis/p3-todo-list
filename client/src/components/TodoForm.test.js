import ReactDom from 'react-dom'
import renderer from 'react-test-renderer'
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react'

import TodoForm from './TodoForm'

let onStatusChangeSpy, onSubmitSpy, todoFormComponent

beforeEach(() => {
  onStatusChangeSpy = jest.fn()
  onSubmitSpy = jest.fn().mockResolvedValueOnce(null)

  todoFormComponent = <TodoForm
    status="all"
    onStatusChange={onStatusChangeSpy}
    onSubmit={onSubmitSpy}
  />
})

afterEach(cleanup)

test('renders in the DOM', () => {
  const div = document.createElement('div')

  ReactDom.render(todoFormComponent, div)
})

test('renders form', () => {
  const { getByTestId } = render(todoFormComponent)

  expect(getByTestId('todo-form')).toBeInTheDocument()
})

test('matches snapshot', () => {
  const tree = renderer.create(todoFormComponent).toJSON()

  expect(tree).toMatchSnapshot()
})

test('submits form', async () => {
  const { getByTestId } = render(todoFormComponent)
  const mockDescription = 'A mock todo item'

  fireEvent.change(getByTestId('todo-status-description'), { target: { value: mockDescription } })
  fireEvent.click(getByTestId('todo-form-button'))

  await waitFor(() => {
    expect(getByTestId('todo-status-description')).toHaveValue('')
  })

  expect(onSubmitSpy).toHaveBeenCalledWith(mockDescription)
})
