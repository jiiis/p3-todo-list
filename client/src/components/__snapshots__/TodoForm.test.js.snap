// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <form
    data-testid="todo-form"
    onSubmit={[Function]}
  >
    <label
      htmlFor="todo-description-field"
    >
      Create todo item
    </label>
    <input
      data-testid="todo-status-description"
      id="todo-description-field"
      name="description"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      data-testid="todo-form-button"
    >
      Add
    </button>
  </form>
  <div>
    <label
      htmlFor="todo-status-select"
    >
      Select status
    </label>
    <select
      data-testid="todo-status-select"
      id="todo-status-select"
      name="status"
      onChange={[Function]}
      value="all"
    >
      <option
        value="all"
      >
        All
      </option>
      <option
        value="completed"
      >
        Completed
      </option>
      <option
        value="uncompleted"
      >
        Uncompleted
      </option>
    </select>
  </div>
</div>
`;
